src/python/pants/engine/fs_test.py

```diff
@@ -865,13 +778,116 @@ def mutation_function(project_tree, dir_path):
+    )
+
+    @classmethod
+    def rules(cls):
+        return (
+            *super().rules(),
+            QueryRule(Snapshot, (DownloadFile,)),
+        )
+
+    def test_download(self) -> None:
+        with self.isolated_local_store():
+            with http_server(StubHandler) as port:
+                snapshot = self.request_product(
+                    Snapshot,
+                    [DownloadFile(f"http://localhost:{port}/file.txt", self.file_digest)],
+                )
+                self.assert_snapshot_equals(
+                    snapshot,
+                    ["file.txt"],
+                    self.expected_snapshot_digest,
+                )
+
+    def test_download_missing_file(self) -> None:
+        with self.isolated_local_store():
+            with http_server(StubHandler) as port:
+                with self.assertRaises(ExecutionError) as cm:
+                    self.request_product(
+                        Snapshot,
+                        [DownloadFile(f"http://localhost:{port}/notfound", self.file_digest)],
+                    )
+                assert "404" in str(cm.exception)
+
+    def test_download_wrong_digest(self) -> None:
+        with self.isolated_local_store():
+            with http_server(StubHandler) as port:
+                with self.assertRaises(ExecutionError) as cm:
+                    self.request_product(
+                        Snapshot,
+                        [
+                            DownloadFile(
+                                f"http://localhost:{port}/file.txt",
+                                Digest(
+                                    self.file_digest.fingerprint,
+                                    self.file_digest.serialized_bytes_length + 1,
+                                ),
+                            )
+                        ],
+                    )
+                assert "wrong digest" in str(cm.exception).lower()
+
+    def test_caches_downloads(self) -> None:
+        with self.isolated_local_store():
+            with http_server(StubHandler) as port:
+                self.prime_store_with_roland_digest()
 
-See src/python/pants/engine/fs_test.py in the pants repo for details.
-"""
+                # This would error if we hit the HTTP server, because 404,
+                # but we're not going to hit the HTTP server because it's cached,
+                # so we shouldn't see an error...
+                url = DownloadFile(
+                    f"http://localhost:{port}/roland",
+                    Digest("693d8db7b05e99c6b7a7c0616456039d89c555029026936248085193559a0b5d", 16),
+                )
+                snapshot = self.request_product(Snapshot, [url])
+                self.assert_snapshot_equals(
+                    snapshot,
+                    ["roland"],
+                    Digest("9341f76bef74170bedffe51e4f2e233f61786b7752d21c2339f8ee6070eba819", 82),
+                )
+
+    def test_download_https(self) -> None:
```

Comment:
This is the relevant new test, the other methods in this class are copied (with one or two tiny mods).